From af1c8534f29178489c430a16dc668dd79277e4f8 Mon Sep 17 00:00:00 2001
From: David Ward <david.ward@ll.mit.edu>
Date: Thu, 14 May 2026 07:07:30 -0400
Subject: [PATCH 5/5] lib: northbound: Replace use of lyd_new_term_bin()

lyd_new_term_bin() was removed in libyang 5.1.3. Adjust the get()
operational callbacks in FRR that used this function.

Signed-off-by: David Ward <david.ward@ll.mit.edu>
Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>
---
 lib/if.c                              | 80 ++++-----------------------
 lib/mgmt_be_client.c                  | 18 +++---
 lib/yang.c                            |  8 ---
 lib/yang.h                            |  7 ---
 tests/lib/northbound/test_oper_data.c |  7 +--
 5 files changed, 23 insertions(+), 97 deletions(-)

diff --git a/lib/if.c b/lib/if.c
index 2791481be6..04570114bc 100644
--- a/lib/if.c
+++ b/lib/if.c
@@ -1762,71 +1762,6 @@ static enum nb_error lib_interface_vrf_get(const struct nb_node *nb_node, const
 	return NB_OK;
 }
 
-/*
- * XPath: /frr-interface:lib/interface/state/if-index
- */
-static enum nb_error lib_interface_state_if_index_get(const struct nb_node *nb_node,
-						      const void *list_entry,
-						      struct lyd_node *parent)
-{
-	const struct lysc_node *snode = nb_node->snode;
-	const struct interface *ifp = list_entry;
-	int32_t value = ifp->ifindex;
-
-	if (yang_new_term_bin(parent, snode->module, snode->name, &value, sizeof(value),
-			      LYD_NEW_PATH_UPDATE, NULL))
-		return NB_ERR_RESOURCE;
-	return NB_OK;
-}
-
-/*
- * XPath: /frr-interface:lib/interface/state/mtu[6]
- */
-static enum nb_error lib_interface_state_mtu_get(const struct nb_node *nb_node,
-						 const void *list_entry, struct lyd_node *parent)
-{
-	const struct lysc_node *snode = nb_node->snode;
-	const struct interface *ifp = list_entry;
-	uint32_t value = ifp->mtu;
-
-	if (yang_new_term_bin(parent, snode->module, snode->name, &value, sizeof(value),
-			      LYD_NEW_PATH_UPDATE, NULL))
-		return NB_ERR_RESOURCE;
-	return NB_OK;
-}
-
-/*
- * XPath: /frr-interface:lib/interface/state/speed
- */
-static enum nb_error lib_interface_state_speed_get(const struct nb_node *nb_node,
-						   const void *list_entry, struct lyd_node *parent)
-{
-	const struct lysc_node *snode = nb_node->snode;
-	const struct interface *ifp = list_entry;
-	uint32_t value = ifp->speed;
-
-	if (yang_new_term_bin(parent, snode->module, snode->name, &value, sizeof(value),
-			      LYD_NEW_PATH_UPDATE, NULL))
-		return NB_ERR_RESOURCE;
-	return NB_OK;
-}
-
-/*
- * XPath: /frr-interface:lib/interface/state/metric
- */
-static enum nb_error lib_interface_state_metric_get(const struct nb_node *nb_node,
-						    const void *list_entry, struct lyd_node *parent)
-{
-	const struct lysc_node *snode = nb_node->snode;
-	const struct interface *ifp = list_entry;
-	uint32_t value = ifp->metric;
-
-	if (yang_new_term_bin(parent, snode->module, snode->name, &value, sizeof(value),
-			      LYD_NEW_PATH_UPDATE, NULL))
-		return NB_ERR_RESOURCE;
-	return NB_OK;
-}
-
 /*
  * XPath: /frr-interface:lib/interface/state/phy-address
  */
@@ -1877,31 +1812,36 @@ const struct frr_yang_module_info frr_interface_info = {
 		{
 			.xpath = "/frr-interface:lib/interface/state/if-index",
 			.cbs = {
-				.get = lib_interface_state_if_index_get,
+				.get = nb_oper_int32_get,
+				.get_elem = (void *)(intptr_t)offsetof(struct interface, ifindex),
 			}
 		},
 		{
 			.xpath = "/frr-interface:lib/interface/state/mtu",
 			.cbs = {
-				.get = lib_interface_state_mtu_get,
+				.get = nb_oper_uint32_get,
+				.get_elem = (void *)(intptr_t)offsetof(struct interface, mtu),
 			}
 		},
 		{
 			.xpath = "/frr-interface:lib/interface/state/mtu6",
 			.cbs = {
-				.get = lib_interface_state_mtu_get,
+				.get = nb_oper_uint32_get,
+				.get_elem = (void *)(intptr_t)offsetof(struct interface, mtu),
 			}
 		},
 		{
 			.xpath = "/frr-interface:lib/interface/state/speed",
 			.cbs = {
-				.get = lib_interface_state_speed_get,
+				.get = nb_oper_uint32_get,
+				.get_elem = (void *)(intptr_t)offsetof(struct interface, speed),
 			}
 		},
 		{
 			.xpath = "/frr-interface:lib/interface/state/metric",
 			.cbs = {
-				.get = lib_interface_state_metric_get,
+				.get = nb_oper_uint32_get,
+				.get_elem = (void *)(intptr_t)offsetof(struct interface, metric),
 			}
 		},
 		{
diff --git a/lib/mgmt_be_client.c b/lib/mgmt_be_client.c
index df9cc471d4..2a7654e32f 100644
--- a/lib/mgmt_be_client.c
+++ b/lib/mgmt_be_client.c
@@ -1244,15 +1244,16 @@ static enum nb_error clients_client_state_candidate_config_version_get(
 	const struct nb_node *nb_node, const void *parent_list_entry, struct lyd_node *parent)
 {
 	const struct lysc_node *snode = nb_node->snode;
-	uint64_t value;
+	uint64_t version;
+	char value[21]; /* UINT64_MAX has 20 digits */
 
 	if (__be_client->config_txn)
-		value = __be_client->config_txn->candidate_config->version;
+		version = __be_client->config_txn->candidate_config->version;
 	else
-		value = __be_client->running_config->version;
+		version = __be_client->running_config->version;
 
-	if (yang_new_term_bin(parent, snode->module, snode->name, &value, sizeof(value),
-			      LYD_NEW_PATH_UPDATE, NULL))
+	snprintfrr(value, sizeof(value), "%" PRIu64, version);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
 		return NB_ERR_RESOURCE;
 
 	return NB_OK;
@@ -1266,10 +1267,11 @@ static enum nb_error clients_client_state_running_config_version_get(const struc
 								     struct lyd_node *parent)
 {
 	const struct lysc_node *snode = nb_node->snode;
-	uint64_t value = __be_client->running_config->version;
+	uint64_t version = __be_client->running_config->version;
+	char value[21]; /* UINT64_MAX has 20 digits */
 
-	if (yang_new_term_bin(parent, snode->module, snode->name, &value, sizeof(value),
-			      LYD_NEW_PATH_UPDATE, NULL))
+	snprintfrr(value, sizeof(value), "%" PRIu64, version);
+	if (lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL))
 		return NB_ERR_RESOURCE;
 
 	return NB_OK;
diff --git a/lib/yang.c b/lib/yang.c
index 40dce87b35..15533d3ec2 100644
--- a/lib/yang.c
+++ b/lib/yang.c
@@ -1715,7 +1715,6 @@ LY_ERR yang_lyd_parse_data(const struct ly_ctx *ctx, struct lyd_node *parent,
  * Handle NBC API changes between versions of Libyang
  */
 
-#undef lyd_new_term_bin
 #undef lyd_new_path2
 
 #if (LY_VERSION_MAJOR >= 4)
@@ -1724,13 +1723,6 @@ LY_ERR yang_lyd_parse_data(const struct ly_ctx *ctx, struct lyd_node *parent,
 #define LY_SZ(x) (x)
 #endif
 
-LY_ERR yang_new_term_bin(struct lyd_node *parent, const struct lys_module *module,
-			 const char *name, const void *value, uint32_t size, uint32_t options,
-			 struct lyd_node **node)
-{
-	return lyd_new_term_bin(parent, module, name, value, LY_SZ(size), options, node);
-}
-
 LY_ERR yang_new_path2(struct lyd_node *parent, const struct ly_ctx *ctx, const char *path,
 		      const void *value, uint32_t value_size_bytes, uint32_t options,
 		      struct lyd_node **new_parent, struct lyd_node **new_node)
diff --git a/lib/yang.h b/lib/yang.h
index dac6c43239..5487327ee2 100644
--- a/lib/yang.h
+++ b/lib/yang.h
@@ -950,13 +950,6 @@ extern LY_ERR yang_lyd_parse_data(const struct ly_ctx *ctx,
  * Handle NBC API changes between versions of Libyang
  */
 
-/* Use this helper function to deal with change in value_size units */
-extern LY_ERR yang_new_term_bin(struct lyd_node *parent, const struct lys_module *module,
-				const char *name, const void *value, uint32_t value_size,
-				uint32_t options, struct lyd_node **node);
-#define lyd_new_term_bin(...)                                                                     \
-	_Pragma("GCC error \"Use yang_new_term_bin() instead of lyd_new_term_bin()\"")
-
 extern LY_ERR yang_new_path2(struct lyd_node *parent, const struct ly_ctx *ctx, const char *path,
 			     const void *value, uint32_t value_size_bytes, uint32_t options,
 			     struct lyd_node **new_parent, struct lyd_node **new_node);
diff --git a/tests/lib/northbound/test_oper_data.c b/tests/lib/northbound/test_oper_data.c
index c16cecb198..e7756a4c33 100644
--- a/tests/lib/northbound/test_oper_data.c
+++ b/tests/lib/northbound/test_oper_data.c
@@ -249,12 +249,11 @@ static enum nb_error frr_test_module_c2cont_c2value_get(const struct nb_node *nb
 							struct lyd_node *parent)
 {
 	const struct lysc_node *snode = nb_node->snode;
-	uint32_t value = htole32(0xAB010203);
+	char value[11]; /* UINT32_MAX has 10 digits */
 	LY_ERR err;
 
-	/* Note that this api expects 'value' to be in little-endian form */
-	err = yang_new_term_bin(parent, snode->module, snode->name, &value, sizeof(value),
-				LYD_NEW_PATH_UPDATE, NULL);
+	snprintfrr(value, sizeof(value), "%" PRIu32, 0xAB010203);
+	err = lyd_new_term(parent, snode->module, snode->name, value, LYD_NEW_PATH_UPDATE, NULL);
 	assert(err == LY_SUCCESS);
 
 	return NB_OK;
-- 
2.55.0

